### ISSUE TITLE
Reloading TESGroup uses old filters

### ISSUE THREAD

**Issue by joefowler** [usnistgov/mass#82]
**[Original report](https://bitbucket.org/joe_fowler/mass/issues/82) by Charles James Titus (Bitbucket: [cjtitus](https://bitbucket.org/cjtitus), GitHub: [cjtitus](https://github.com/cjtitus)).**

----------------------------------------

tldr; If you filter data with the new filters, exit your program, and then reload the data, your datasets will now magically have the old filters, and this breaks things.

Longer explanation, because Galen will know exactly where to test this because I'm using code he sent me:

I have been trying to run Galen's calibration code that transfers filters from a drift correct data set to a calibration data set. The learn_drift_correction.py computes filters using the ArrivalTimeSafeFilter (and then exits).

I.e., you have some TESGroup instance named 'data_dc', and you can run
```
#!python
In[1]: data_dc.channel[1].filter
Out[1]: <mass.core.optimal_filtering.ArrivalTimeSafeFilter at ...>
```
then learn_calibration.py tries to copy those filters to another data set. But when you try to reconstitute your data

```
#!python
In[1]: data_dc = mass.TESGroup(pulse_files_dc, noise_files_dc)
```
you can try

```
#!python
In[2]: data_dc.channel[1].filter 
Out[2]: <mass.core.optimal_filtering.Filter at ...>

```

and see that you don't have the new filters anymore. So as soon as something tries to call filter.aterms, you crash

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

How time critical is this for you? Today vs Monday?

**Comment by joefowler** [comment]
**Original comment by Charles James Titus (Bitbucket: [cjtitus](https://bitbucket.org/cjtitus), GitHub: [cjtitus](https://github.com/cjtitus)).**

----------------------------------------

I can work around it for now. If I just smash all the programs together and never reload the data it's ok. 

Also haven't tried running the development branch, I'm on Master. Did this work on the dev branch?

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

I probably didn't test it as thoroughly as you are testing it. I think the hdf5 file for `data_dc` existed before I ran `learn_drift_correct` and had old filters in it, so I was always "reloading".

**Comment by joefowler** [comment]
**Original comment by Charles James Titus (Bitbucket: [cjtitus](https://bitbucket.org/cjtitus), GitHub: [cjtitus](https://github.com/cjtitus)).**

----------------------------------------

The guess there are two separate problems here:

1.  When loading the hdf5 file from disk, Mass always chooses to use the old filters, even if the hdf5 file was saved with new filters
2.  When you copy the filters from the data_dc to the data, data thinks it's getting new filters, so when you copy old filters over it doesn't bother checking at all, and just fails as soon as it tries to call something that's only in the new filters

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

Filters are created from the hdf5 file here: https://bitbucket.org/joe_fowler/mass/src/9cafef6af578b59e24430fbf8a51ece70c85a4df/src/mass/core/channel.py?at=master&fileviewer=file-view-default#channel.py-654

But I haven't found where filters are written to the HDF5 file. We would need to add some flag for type of filter, then in the loading code check that flag and chose the correct filter constructor and set the `_use_new_filters` variable to the correct value.

edit: To be clear, I'm just trying to write down useful notes. I don't expect you to do this Jamie.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Change how new and old filters are loaded from HDF5.

Now the store-procedure explicitly specifies whether the
filter is old- or new-style. More importantly the loading
procedure checks and creates the appropriate style. Fixes #82.
There's also a test_filtering.py which tests this fix.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Jamie: for now this fix is only in branch:develop. Let me know if that's highly inconvenient.
